src/main/webapp/app/shared/user-settings/ide-preferences/ide-settings.component.ts

```diff
@@ -0,0 +1,80 @@
+
+    constructor(private ideSettingsService: IdeSettingsService) {}
+
+    ngOnInit() {
+        this.ideSettingsService.loadPredefinedIdes().subscribe((predefinedIde) => {
+            this.PREDEFINED_IDE = predefinedIde;
+        });
+
+        this.ideSettingsService.loadIdePreferences().subscribe((programmingLanguageToIdeMap) => {
+            if (!programmingLanguageToIdeMap.has(ProgrammingLanguage.EMPTY)) {
+                programmingLanguageToIdeMap.set(ProgrammingLanguage.EMPTY, this.PREDEFINED_IDE[0]);
+            }
+
+            this.programmingLanguageToIde.set(programmingLanguageToIdeMap);
+
+            // initialize assigned programming languages
+            this.assignedProgrammingLanguages = Array.from(programmingLanguageToIdeMap.keys()).filter((x: ProgrammingLanguage) => x !== ProgrammingLanguage.EMPTY);
+
+            // initialize remaining programming languages
+            this.remainingProgrammingLanguages = Array.from(
+                Object.values(ProgrammingLanguage).filter((x) => !this.assignedProgrammingLanguages.includes(x) && x !== ProgrammingLanguage.EMPTY),
+            );
+        });
+    }
+
+    addProgrammingLanguage(programmingLanguage: ProgrammingLanguage) {
+        this.ideSettingsService.saveIdePreference(programmingLanguage, this.PREDEFINED_IDE[0]).subscribe((ide) => {
+            this.programmingLanguageToIde.update((map) => new Map(map.set(programmingLanguage, ide)));
+
+            this.assignedProgrammingLanguages.push(programmingLanguage);
+            this.remainingProgrammingLanguages = this.remainingProgrammingLanguages.filter((x) => x !== programmingLanguage);
+        });
+    }
+
+    changeIde(programmingLanguage: ProgrammingLanguage, ide: Ide) {
+        this.ideSettingsService.saveIdePreference(programmingLanguage, ide).subscribe((ide) => {
+            this.programmingLanguageToIde.update((map) => new Map(map.set(programmingLanguage, ide)));
+        });
+    }
+
+    removeProgrammingLanguage(programmingLanguage: ProgrammingLanguage) {
+        this.ideSettingsService.deleteIdePreference(programmingLanguage).subscribe(() => {
+            const programmingLanguageToIdeMap: Map<ProgrammingLanguage, Ide> = new Map(this.programmingLanguageToIde());
+            programmingLanguageToIdeMap.delete(programmingLanguage);
+
+            this.programmingLanguageToIde.set(programmingLanguageToIdeMap);
+
+            this.remainingProgrammingLanguages.push(programmingLanguage);
+            this.assignedProgrammingLanguages = this.assignedProgrammingLanguages.filter((x) => x !== programmingLanguage);
+        });
+    }
+
+    isIdeOfProgrammingLanguage(programmingLanguage: ProgrammingLanguage, ide: Ide): boolean {
+        return ideEquals(this.programmingLanguageToIde().get(programmingLanguage), ide);
+    }
+
+    protected readonly ProgrammingLanguage = ProgrammingLanguage;
+    protected readonly faPlus = faPlus;
+    protected readonly faTrash = faTrash;
```

Comment:
The `protected PREDEFINED_IDE` is at the top of the file. I think it makes sense to move these fields there as well

Reply:
done